### PR TITLE
Implement SEP-990 Enterprise Managed OAuth

### PR DESCRIPTION
**Pull request by sagar-okta**
This PR implements SEP-990 which adds support for Enterprise Managed OAuth using RFC 8693 Token Exchange and RFC 7523 JWT Bearer flows. This enables secure machine-to-machine authentication for MCP clients in enterprise environments without requiring user interaction.

Related: #1090

## Motivation and Context
Enterprise environments often require more secure OAuth flows that don't involve user interaction for machine-to-machine communication. SEP-990 addresses this by implementing:

Token Exchange (RFC 8693): Allows exchanging an ID token from an enterprise IDP for an authorization grant
JWT Bearer Grant (RFC 7523): Enables exchanging the authorization grant for an access token to access MCP resources
This change is needed to support enterprise customers who need to integrate MCP clients into their existing OAuth infrastructure securely.

## How Has This Been Tested?
Added comprehensive unit tests in xaa-util.test.ts (994 new test cases) covering:
Successful token exchange flows
Authorization grant request failures (400, 401, 500 errors)
Access token exchange failures
OAuth error handling (invalid_request, invalid_client, invalid_grant, etc.)
Edge cases and validation (empty responses, malformed JSON, special characters encoding)
Token type validation
Request body encoding
Added middleware tests in middleware.test.ts (55 additional test cases)
Added documentation in [client.md](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)

## Breaking Changes
No breaking changes - This is an additive feature that introduces new functionality without modifying existing APIs.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
Implementation Details:

New utility module xaa-util.ts (593 lines) implementing the core token exchange logic
New middleware functions in middleware.ts for integrating XAA into the auth flow
Added qs dependency for proper URL encoding of OAuth request parameters
Comprehensive error handling for various OAuth error responses
Support for OAuth metadata discovery for both IDP and MCP authorization servers

### CODE REVIEW

**Comment by pcarleton** [comment]
Hi @sagar-okta, I owe you a review on this but won't be able to get to it until Jan 23rd while I wrap up conformance tests for SDK tiering.

To make progress in the meantime, a conformance test for this feature would be really helpful to ensure the implementations are compatible across SDKs.

Cross-linking: https://github.com/modelcontextprotocol/python-sdk/pull/1721

Thanks for your patience!

File: packages/client/src/client/xaaUtil.ts [hunk 1]
```diff
@@ -0,0 +1,597 @@
+import type { FetchLike } from '@modelcontextprotocol/core';
+import qs from 'qs';
+
+import { discoverAuthorizationServerMetadata } from './auth.js';
+// ============================================================================
+// CONSTANTS
+// ============================================================================
+
+const OAuthErrorTypes = [
```
Comment (pcarleton) [inline packages/client/src/client/xaaUtil.ts]: please use OAuthErrorCode from packages/core/src/auth/errors.ts  and edit there as needed

File: packages/client/src/client/xaaUtil.ts [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+    'unauthorized_client',
+    'unsupported_grant_type',
+    'invalid_scope'
+] as const;
+
+// ============================================================================
+// ENUMS
+// ============================================================================
+
+const enum OAuthGrantType {
```
Comment (pcarleton) [inline packages/client/src/client/xaaUtil.ts]: similar here, please add these to core if they're not already there.

File: packages/client/src/client/xaaUtil.ts [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+    JWT_BEARER = 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer'
+}
+
+// ============================================================================
+// TYPES
+// ============================================================================
+
+type OAuthErrorType = (typeof OAuthErrorTypes)[number];
+
+type OAuthError = {
```
Comment (pcarleton) [inline packages/client/src/client/xaaUtil.ts]: these should also be able to use the types from core.

**Comment by sagar-okta** [comment]
Hi @pcarleton 
The middleware approach enables us to expose the access token retrieval method to anyone who wants to use cross app access. Also it was much more simpler to make any new enhancements in the future if required since we had to only update the method. Since we already had a similar middleware for this we went with this approach. This can be further used by Streamablehttps and others easily by just providing params and a fetch function.

File: packages/client/src/client/xaaUtil.ts (hunk 2 above)
Comment (sagar-okta) [reply packages/client/src/client/xaaUtil.ts]: Moved the enum to auth.ts inside core

File: packages/client/src/client/xaaUtil.ts (hunk 3 above)
Comment (sagar-okta) [reply packages/client/src/client/xaaUtil.ts]: Though a similar type was found in core we kept the existing one since we are using enum for error type. Moved this as well to auth.ts under core.

**Comment by sagar-okta** [comment]
Hi @pcarleton 

I've made the required changes and pushed. Can you please review it?
I had to rebase and force push during that time I missed your commit https://github.com/modelcontextprotocol/typescript-sdk/commit/ec00df3b2058c5ded5b0c985c3bf6fb94b6665a9 please let me know if it is required.

**Comment by pcarleton** [comment]
Hi @sagar-okta thanks for this. 

The commit you linked is a fork of this PR that explored the provider approach to see how it looked.  I do think the Provider approach is a better fit for this change since it sits it more similarly with the client credentials extension.  If you can align this PR with that approach, we can merge it in.

**Comment by sagar-okta** [comment]
Hi @pcarleton , I've made the changes as requested. I've created a new PR, so that we can keep this if required later.
I'll be closing this PR as of now. Please review the below pr attached.
New PR : https://github.com/modelcontextprotocol/typescript-sdk/pull/1593
